pgx-pg-sys/src/lib.rs

```diff
@@ -445,6 +450,42 @@ mod all_versions {
             context: *mut ::std::os::raw::c_void,
         ) -> bool;
     }
+
+    #[inline]
+    pub fn get_arr_data_ptr<T>(arr: *mut super::ArrayType) -> *mut T {
+        unsafe { pgx_ARR_DATA_PTR(arr) as *mut T }
+    }
+
+    #[inline]
+    pub fn get_arr_nelems(arr: *mut super::ArrayType) -> i32 {
+        unsafe { pgx_ARR_NELEMS(arr) }
+    }
+
+    #[inline]
+    pub fn get_arr_ndim(arr: *mut super::ArrayType) -> i32 {
+        unsafe { pgx_ARR_NDIM(arr) }
+    }
+
+    #[inline]
+    pub fn get_arr_nullbitmap<'a>(arr: *mut super::ArrayType) -> &'a [pg_sys::bits8] {
+        unsafe {
+            let len = (pgx_ARR_NELEMS(arr) + 7) / 8;
+            std::slice::from_raw_parts(pgx_ARR_NULLBITMAP(arr), len as usize)
+        }
+    }
+
+    #[inline]
+    pub fn get_arr_nullbitmap_mut<'a>(arr: *mut super::ArrayType) -> &'a mut [u8] {
+        unsafe {
+            let len = (pgx_ARR_NELEMS(arr) + 7) / 8;
+            std::slice::from_raw_parts_mut(pgx_ARR_NULLBITMAP(arr), len as usize)
+        }
+    }
+
+    #[inline]
+    pub fn get_arr_hasnull(arr: *mut super::ArrayType) -> bool {
+        unsafe { pgx_ARR_HASNULL(arr) }
+    }
```

Comment:
I believe none of these are actually safe due to the prior reason: a field read on a possibly-invalid, possibly-null pointer happens.

Reply:
I have introduced safe wrappers for everything except the null bitmap functions, which will require another wave of commits.